opm/simulators/linalg/getQuasiImpesWeights.hpp

```diff
@@ -87,6 +87,95 @@ namespace Amg
+        constexpr int numEq = VectorBlockType::size();
+        using Evaluation = typename std::decay_t<decltype(model.localLinearizer(threadId).localResidual().residual(0))>
+            ::block_type;
+        using Toolbox = MathToolbox<Evaluation>;
+        VectorBlockType rhs(0.0);
+        const auto& solution= model.solution(/*timeIdx*/0);
+        auto elemIt = gridView.template begin</*codim=*/0>();
+        const auto& elemEndIt = gridView.template end</*codim=*/0>();
+        for (; elemIt != elemEndIt; ++elemIt) {
+            elemCtx.updatePrimaryStencil(*elemIt);
+            elemCtx.updatePrimaryIntensiveQuantities(/*timeIdx=*/0);
+            const auto& index = elemCtx.globalSpaceIndex(/*spaceIdx=*/0, /*timeIdx=*/0);
+            const auto& intQuants = elemCtx.intensiveQuantities(/*spaceIdx=*/0, /*timeIdx=*/0);
+            const auto& fs = intQuants.fluidState();
+            VectorBlockType bweights;
+           
+            
+            if(FluidSystem::phaseIsActive(FluidSystem::waterPhaseIdx)){
+                unsigned activeCompIdx = Indices::canonicalToActiveComponentIndex(
+                    FluidSystem::solventComponentIndex(FluidSystem::waterCompIdx)
+                    );
+                bweights[FluidSystem::waterPhaseIdx]= Toolbox::template decay<LhsEval>(1/fs.invB(FluidSystem::waterPhaseIdx));
+            }
+            double denominator=1.0;
+            if (FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx) &&
+                FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx) ){
+                denominator =  Toolbox::template decay<LhsEval>(1 - fs.Rs()*fs.Rv());
+            }
+            
+            if( not(denominator>0)){
+                std::cout << "Probably negative compressibility" << std::endl;
+            }
+            if( FluidSystem::phaseIsActive(FluidSystem::oilPhaseIdx) ){
+                unsigned activeCompIdx = Indices::canonicalToActiveComponentIndex(
+                    FluidSystem::solventComponentIndex(FluidSystem::oilCompIdx)
+                    );
+                bweights[activeCompIdx]= Toolbox::template decay<LhsEval>(
+                    (1/fs.invB(FluidSystem::oilPhaseIdx)-fs.Rs()
+                    /fs.invB(FluidSystem::gasPhaseIdx))/denominator);
+            }
+            if(FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)){
+                 unsigned activeCompIdx = Indices::canonicalToActiveComponentIndex(
+                        FluidSystem::solventComponentIndex(FluidSystem::gasCompIdx)
+                        );
+                 bweights[activeCompIdx]= Toolbox::template decay<LhsEval>(
+                     (1/fs.invB(FluidSystem::gasPhaseIdx)-fs.Rv()
+                      /fs.invB(FluidSystem::oilPhaseIdx))/denominator);
+            }
+            const auto& meaning =  solution[index].primaryVarsMeaning();
+            if( (meaning == PrimaryVariables::Sw_pg_Rv) && (meaning == PrimaryVariables::Sw_po_Rs)){
```

Comment:
Can it really have two different meanings at the same time?

Reply:
It may be an error, but it may also be correct to skipt the full if sentence since bweighs probably is zero in this case anyway. I have changed to the code is clearer.

Reply:
also @blattms 
 could you find the correct transpose from dune. Here it is hardcoded implementation in top of the file.

Reply:
Unfortunately this depends on the version of DUNE.
+ <=2.8 only has method [`transpose`](https://dune-project.org/doxygen/2.7.0/classDune_1_1Matrix.html#a29efbcc869ca7da30095c9004f3ecbc1) for dense matrix returning the transposed. Not sure whether this is a performance problem. Maybe benchmark?
+ master has an additional wrapper for the transpose and free functions to get it see  [here](https://dune-project.org/doxygen/master/namespaceDune.html#ad090bea53c1598805a721f989ffea309)